File: Delegate/DelegateInObjectiveC/Delegate/PatternProvider.m
//
//  PatternProvider.m
//  Delegate
//
//  Created by Ivaldo De Sousa on 2018-10-01.
//  Copyright © 2018 Ivaldo De Sousa. All rights reserved.
//

#import "PatternProvider.h"

@implementation PatternProvider
- (void)draw
{
    int space = _delegate.spaceBetweenCharacter;
    int emptyLine = _delegate.emptyLineBetweenLines;
    char character = _delegate.characterFunction;
    int lines = _delegate.numberOfLines;
    NSString* pattern = _delegate.patternType;
    
    
    if ([pattern  isEqual: @"Triangle"]){
        for (int i=0;i<lines;i++){
            NSLog(@"%c",character);
            if (emptyLine>0){
                for (int i=0;i<emptyLine;i++){
                    NSLog(@"%s"," ");
                }
            }
        }
        
    }
    if ([pattern  isEqual: @"ReversedTriangle"]){
        for (int i=0;i<lines;i++){
            NSLog(@"%c",character);
            if (emptyLine>0){
                for (int i=0;i<emptyLine;i++){
                    NSLog(@"%s"," ");
                }
            }
        }
        
    }
    
    

}
@end

File: Delegate/DelegationInJava/src/PatternProvider.java


public class PatternProvider{
	
	private DelegateInterface delegate;
	

	public DelegateInterface getDelegate() {
		return delegate;
	}

	public void setDelegate(DelegateInterface delegate) {
		this.delegate = delegate;
	}
	
	private String turnToString() {
		int nSpace = delegate.numberOfSpaceBetweenCharacter();
		char character = delegate.character();
		String newString = Character.toString(character);
		if (nSpace>0) {
			for(int i=0;i<nSpace;i++) {
				newString = newString + " ";
			}
		}
		
		return newString;
	}
	
	
	private void printStar(int n){
	    if(n > 0){
	        System.out.print(turnToString());
	        printStar(n-1);
	    }
	}
	
	public void draw() {
		int nlines = delegate.numberOfEmptyLineBetweenlines();
		int numberOfLines = delegate.numberOfLines();
		String pattern = delegate.pattern();
		
		
		
		if (pattern == "Triangle") {
			for(int i=1; i<numberOfLines;i++) {
				printStar(i);
				for(int j=0;j<nlines;j++ ) {
					System.out.print("\n");
				}
				
				}
		}
		
		if (pattern == "ReversedTriangle") {
			for(int i=numberOfLines; i>0;i--) {
				printStar(i);
				for(int j=0;j<=nlines;j++ ) {
					System.out.print("\n");
				}
				}
		}
		
	}
	
}
